### PR TITLE
Added ElementAverageFunctionDifference

### LINKED ISSUE #14800: ElementFunctionDifferencePostprocessor
## Reason
It is sometimes helpful to compare an elemental variable (e.g. material property output) to a function (e.g. analytical expression).

## Design
This can be done now by using several postprocessors:
```
  [./x]
    type = ElementAverageValue
    variable = x
  [../]
  [./x_exact]
    type = FunctionValuePostprocessor
    function = x_exact
  [../]
  [./x_diff]
    type = DifferencePostprocessor
    value1 = x
    value2 = x_exact
  [../]
```
This can be condensed into one postprocessor

```
[./x_diff]
  type = ElementFunctionDifferencePostprocessor
  variable = x
  function = x_exact
[../]
```

## Impact
Easier hand-calculation comparisons.

### PR DESCRIPTION
**Pull request by tophmatthews**
Closes #14800

### CODE REVIEW

File: framework/src/postprocessors/ElementAverageFunctionDifference.C [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    _abs(getParam<bool>("absolute_value"))
+{
+}
+
+Real
+ElementAverageFunctionDifference::getValue()
+{
+  const Real integral = ElementAverageValue::getValue();
+
+  gatherSum(_volume);
```
Comment (loganharbour) [inline framework/src/postprocessors/ElementAverageFunctionDifference.C]: I believe you're grabbing the sum of `_volume` twice here as `ElementAverageValue::getValue()` gathers the same sum.

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: framework/src/postprocessors/ElementAverageFunctionDifference.C (hunk 1 above)
Comment (loganharbour) [reply framework/src/postprocessors/ElementAverageFunctionDifference.C]: Actually... line 42 here is just the average value. Think this should all just be

```
const Real val = ElementAverageValue::getValue();
const Real diff = avg - _func.value(_t, _q_point[_qp])
```

...right?

Comment (tophmatthews) [reply framework/src/postprocessors/ElementAverageFunctionDifference.C]: yeah, totally! good catch.

**Comment by loganharbour** [comment]
Are you trying to take the volume average of the function too? Because I think you should probably also do the integral over the function and divide by volume

**Comment by tophmatthews** [comment]
> Are you trying to take the volume average of the function too?

No, I want the absolute value of the function, e.g. ParsedFunction

**Comment by loganharbour** [comment]
> > Are you trying to take the volume average of the function too?
> 
> No, I want the absolute value of the function, e.g. ParsedFunction

Okay, but `_func.value(_t, _q_point[_qp])` is going to be evaluated at `_qp = _q_rule->n_points() - 1` (as determined by the end of the `computeIntegral()` loop which seems arbitrary. 

I suppose I'm confused as to why you would want to integrate the variable value so that you can get the average but then only take a single arbitrary point value of the function. You would at least want to evaluate the function at the centroid, yes?

**Comment by tophmatthews** [comment]
Check out #14800, basically I want to replace

```
  [./x]
    type = ElementAverageValue
    variable = x
  [../]
  [./x_exact]
    type = FunctionValuePostprocessor
    function = x_exact
  [../]
  [./x_diff]
    type = DifferencePostprocessor
    value1 = x
    value2 = x_exact
  [../]
```

I get what you're saying about the point, maybe a fake point to pass would be better, or even have the point as a input parameter like `FunctionValuePostprocessor`.

Under what method should the function so it's only called once?

**Comment by loganharbour** [comment]
Hmm. So - in your case you seek the (average element value over the entire domain) - (some function that is the analytic average value over the domain). Is this correct?

If this is the case, I think it would make the most sense to make the function point a parameter. And it should definitely include a note stating that the function is only evaluated at a single point, because otherwise one might easily think that the function is also evaluated per element and maybe even integrated and averaged against the element basis

**Comment by tophmatthews** [comment]
> Hmm. So - in your case you seek the (average element value over the entire domain) - (some function that is the analytic average value over the domain). Is this correct?

Not entirely. The function is typically not domain specific, i.e. not dependent on x,y,z.

**Comment by tophmatthews** [comment]
>I think it would make the most sense to make the function point a parameter.

This probably the most doable...not sure exactly where it'll be domain specific, but at least the option will be there.

**Comment by tophmatthews** [comment]
I'm not sure where else to do the subtraction besides `getValue`? `Finalize` doesn't return anything. Is there a better place?

**Comment by tophmatthews** [comment]
Can you take a look again @loganharbour?

**Comment by loganharbour** [comment]
I should have gotten to you a lot sooner about this - sorry about that.

My concern is that the use case is too specific. That if we were to have something like this, it would involve integrating the function and not just subtracting the function value at a single point. I fear that your use case is best taken care of with multiple PPs

**Comment by joshuahansel** [comment]
I have to agree with Logan here: I think this is too specific and better accomplished with the combination {`ElementAverageValue`, `FunctionValuePostprocessor`, `DifferencePostprocessor`}. Also, the name seems misleading to me; I would think that it's the element average of a difference of functions.

**Comment by tophmatthews** [comment]
Fair enough, closing.

**Pull request closed without merging**